Sources/Base/ASTContext.swift

```diff
@@ -281,24 +282,59 @@ public class ASTContext {
     }
     return bestCandidate?.candidate
   }
+
+  func conformsToProtocol(_ decl: TypeDecl, _ proto: ProtocolDecl) -> Bool {
+    return missingMethodsForConformance(decl, to: proto).isEmpty
+  }
+
+  func haveEqualSignatures(_ decl: FuncDecl, _ other: FuncDecl) -> Bool {
+    guard decl.args.count == other.args.count else { return false }
+    guard decl.hasVarArgs == other.hasVarArgs else { return false }
+    for (declArg, otherArg) in zip(decl.args, other.args) {
+      if declArg.isImplicitSelf && otherArg.isImplicitSelf { continue }
+      guard declArg.externalName == otherArg.externalName else { return false }
+      guard matches(declArg.type, otherArg.type) else { return false }
+    }
+    return true
+  }
   
+  func missingMethodsForConformance(_ decl: TypeDecl, to proto: ProtocolDecl) -> [MethodDecl] {
+    guard let methods = requiredMethods(for: proto) else { return [] }
+    var missing = [MethodDecl]()
+    for method in methods {
+      var impl: MethodDecl?
+      for candidate in decl.methods(named: method.name.name) {
+        if haveEqualSignatures(method, candidate) {
+          impl = candidate
+          break
+        }
+      }
+      if let impl = impl {
+        impl.satisfiedProtocols.insert(proto)
+      } else {
+        missing.append(method)
+      }
+    }
+    return missing
+  }
+
   /// - Returns: Whether the expression's type was changed
   @discardableResult
   func propagateContextualType(_ contextualType: DataType, to expr: Expr) -> Bool {
     let canTy = canonicalType(contextualType)
     switch expr.semanticsProvidingExpr {
     case let expr as NumExpr:
       if case .int = canTy {
-        expr.type = contextualType
+        expr.type = canTy
```

Comment:
Ideally we propagate nothing and this entire method goes the way of the dinosaurs.